pandas/core/groupby/groupby.py

```diff
@@ -1904,8 +1904,7 @@ def _agg_py_fallback(
             ser = Series(values, copy=False)
         else:
             # We only get here with values.dtype == object
-            # TODO: special case not needed with ArrayManager
-            df = DataFrame(values.T)
+            df = DataFrame(values.T, dtype=values.dtype)
```

Comment:
Perhaps slightly out of scope, but I think it'd be good to add an assert on values.dtype == object here

Reply:
Yeah I'd rather keep this out of here